### PR TITLE
Handle 429 With BackOff

### PR DESCRIPTION
**Pull request by tonysun83**
When we encounter a 429, we retry with a different set of retries and
timeout. This will theoretically reduce client replication overload.
When 429s have stopped, it's possible that a 500 error could occur.
Then the retry mechanism should go back to the original way for
backwards compatibility.

Moving this from https://github.com/apache/couchdb-couch-replicator/pull/40 since ASF is currently in a code freeze.

Tested with https://github.com/cloudant/dyno  and also
https://github.com/cloudant/testy/blob/master/suites/replication/test_continuous_replication.py

against a target ccm server

### CODE REVIEW

**Comment by sagelywizard** [comment]
LGTM

**Comment by rnewson** [comment]
-1.

You've changed how maybe_retry works in the face of error such that it can retry more quickly than it did before (from timer:sleep(erlang:min(Wait, ?MAX_WAIT)) to random:uniform(Wait2)). I think it would be better and clearer if you add a new field to httpdb{}.

For errors, we want to back off more aggressively (multiple seconds and higher). For rate-limiting, we want to back off to the point where 429's stop happening, but no further. We anticipate rate limits in the 10's or 100's of requests per second, so a scheme that backs off rapidly to over 2 hours between requests is going too far. I do think it's right to use random:uniform for the delay for 429's, but I wonder, on reflection, if exponentially expanding that range is the right approach? The scheme also never forgives, the backoff period never reduces, even if no 429's would result. I appreciate this is a very tricky area and we should choose our battles. It would be enough to back off less aggressively here. We can always enhance later (the 429 response might contain useful information about when we should retry or what request rate is acceptable, say).

In summary, at least use a different field so we can leave the established timeout/wait behaviour for errors untouched but also consider a slower ramp up in the back off.

**Comment by tonysun83** [comment]
@rnewson: the idea to add random for maybe_retry errors was due to the belief that additional jitter would benefit overall performance of maybe_retry. This can be part of a separate PR though.

Agreed with a new field in httpdb{} record, that would help the issue where we don't tangle wait times for normal errors and backoff.

@sagelywizard @nickva As for the general backoff algorithm, what about modifying the ramp up base to be proportional to a snapshot of the current number of replications (how can we get that? _active_tasks?). @rnewson: are you ok with the maximum of 2 hours? or should we lower that also?

**Comment by nickva** [comment]
@tonysun83 If you want to think about scaling based on number of replications you can try this approach from another commit: https://github.com/cloudant/couch_replicator/commit/08ed2884c5839ef723255575ce982f7d9d7b689e (basically `RepCount = ets:info(?REP_TO_STATE, size)`.  That is on one node only, in total in a cluster there would be about `length([node()|nodes()] * RepCount` replications. But if there are 2 clusters and they are rate limited by the same account, this might not work as well...

@rnewson sorry the idea to do a random backoff for errors was my fault. I suggested that, and yeah, it should be a different commit. To avoid making the sleep period being smaller, would use random:random(Wait*2), then expected wait is still Wait. The idea being it would avoid a stampede if there is a temporary authentication failure or timeouts happen for a slow connection to a target, replications won't all retry on the same schedule.

Having a separate field in the record for backoffs makes most sense also the idea that we should ramp down makes sense. We could subtract a little bit out of the wait value on successful requests. I think that's the the additive increase / multiplicative decrease algorithm. TCP uses that algorithm for its window size and it is proven to converge.

https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease

**Comment by rnewson** [comment]
if it could converge on the acceptable rate that would be pretty sweet.

**Comment by tonysun83** [comment]
I also like the idea of subtracting out from the backoff time. For TCP, it's one unit of the pre-defined max size that they add to the window size. Our equivalent is one unit of backoff time we want to decrement.
I'm trying to figure out if there is a mathematical way to calculate this value as well as the backoff base value based on some variable (i.e average rate limit among customers, avg latency per send_req, etc)...

Currently, I'm thinking of keeping the base as 2  (rather than something < 2), and the decrement to be the same as our initial wait time of 250 ms.

**Comment by tonysun83** [comment]
@nickva: @rnewson: I looked through the code for send_req, and it seems to me that the backoff time will be reset to the initial value after each send_req. Only within the send_req call will we increase it exponentially. If this is the case, then there is no need to perform a some sort of decrement/reduce?

As for the ramp up, I changed the base to 1.5 and that reduces the ramp up dramatically. What about making this base configurable?

File: src/couch_replicator_httpc.erl [hunk 1]
```diff
@@ -226,21 +234,45 @@ clean_mailbox(_, Count) when Count > 0 ->
...
     Params) ->
-    Method = string:to_upper(atom_to_list(get_value(method, Params, get))),
-    Url = couch_util:url_strip_password(full_url(HttpDb, Params)),
-    twig:log(notice,"Retrying ~s request to ~s in ~p seconds due to error ~s",
-        [Method, Url, Wait / 1000, error_cause(Error)]),
-    ok = timer:sleep(Wait),
+    log_retry_error(Params, HttpDb, Wait, Error),
+    % This is so that a long backoff time is not used to ensure
+    % backwards compatibility.
+    ok = timer:sleep(erlang:min(Wait, ?MAX_WAIT)),
```
Comment (rnewson) [inline src/couch_replicator_httpc.erl]: only the extraction of log_retry_error is needed here now that backoff is its own field?

File: include/couch_replicator_api_wrap.hrl [hunk 2]
```diff
@@ -24,7 +24,8 @@
     retries = 10,
     wait = 250,         % milliseconds
     httpc_pool = nil,
-    http_connections
+    http_connections,
+    backoff = 250
```
Comment (rnewson) [inline include/couch_replicator_api_wrap.hrl]: start lower? we're expecting rate limits of 10's or 100's of requests per second. If this were runtime configurable I would worry less. Bidding starts at 25.

**Comment by rnewson** [comment]
+1, squash first pls.

File: src/couch_replicator_httpc.erl [hunk 3]
```diff
@@ -226,21 +234,43 @@ clean_mailbox(_, Count) when Count > 0 ->
     ok.
 
 
+%% For 429 errors, we perform an exponential backoff up to 2.17 hours.
+%% We use Backoff time as a timeout/failure end.
+backoff(Worker, #httpdb{backoff = Backoff} = HttpDb, Params) ->
+    ok = timer:sleep(random:uniform(Backoff)),
+    Backoff2 = round(Backoff*?BACKOFF_EXP),
+    NewBackoff= erlang:min(Backoff2, ?MAX_BACKOFF_WAIT),
```
Comment (sagelywizard) [inline src/couch_replicator_httpc.erl]: Space
